File: rerun-reports/02-python/20260829-032449/results/storage/data-plane/python/streaming/storage-dp-python-blob-event-notifier/python-azure-skills/baseline/generated-code/event_publisher.py
"""Publish custom downstream notifications to an Event Grid topic."""

from __future__ import annotations

import logging
from collections.abc import Iterable
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any
from uuid import uuid4

from azure.core.exceptions import AzureError
from azure.eventgrid import EventGridEvent, EventGridPublisherClient
from azure.eventgrid.aio import EventGridPublisherClient as AsyncEventGridPublisherClient
from azure.identity import DefaultAzureCredential
from azure.identity.aio import DefaultAzureCredential as AsyncDefaultAzureCredential

LOGGER = logging.getLogger(__name__)


@dataclass(frozen=True)
class CustomEvent:
    event_type: str
    data: Any
    subject: str
    data_version: str = "1.0"

    def to_event_grid_event(self) -> EventGridEvent:
        return EventGridEvent(
            id=str(uuid4()),
            event_time=datetime.now(timezone.utc),
            subject=self.subject,
            event_type=self.event_type,
            data=self.data,
            data_version=self.data_version,
        )


def publish_custom_events(
    topic_endpoint: str,
    custom_events: Iterable[CustomEvent],
    *,
    client: Any | None = None,
) -> bool:
    owned_credential = None
    owned_client = None
    if client is None:
        owned_credential = DefaultAzureCredential()
        owned_client = EventGridPublisherClient(topic_endpoint, owned_credential)
        client = owned_client

    events = [event.to_event_grid_event() for event in custom_events]
    try:
        client.send(events)
        return True
    except AzureError:
        LOGGER.exception("Failed to publish %d custom Event Grid event(s)", len(events))
        return False
    finally:
        if owned_client is not None:
            owned_client.close()
        if owned_credential is not None:
            owned_credential.close()


async def publish_custom_events_async(
    topic_endpoint: str,
    custom_events: Iterable[CustomEvent],
    *,
    client: Any | None = None,
) -> bool:
    owned_credential = None
    owned_client = None
    if client is None:
        owned_credential = AsyncDefaultAzureCredential()
        owned_client = AsyncEventGridPublisherClient(topic_endpoint, owned_credential)
        client = owned_client

    events = [event.to_event_grid_event() for event in custom_events]
    try:
        await client.send(events)
        return True
    except AzureError:
        LOGGER.exception("Failed to publish %d custom Event Grid event(s)", len(events))
        return False
    finally:
        if owned_client is not None:
            await owned_client.close()
        if owned_credential is not None:
            await owned_credential.close()

File: rerun-reports/03-java/20260829-075759/results/storage/data-plane/java/streaming/storage-dp-java-blob-event-notifier/java-azure-skills/azure-skill-mcp-microsoft-skill/generated-code/src/main/java/com/example/blobevents/EventPublisher.java
package com.example.blobevents;

import com.azure.core.util.BinaryData;
import com.azure.messaging.eventgrid.EventGridEvent;
import com.azure.messaging.eventgrid.EventGridPublisherClient;
import com.example.blobevents.model.CustomEvent;
import java.util.List;
import java.util.function.Consumer;

public final class EventPublisher {
    private final Consumer<List<EventGridEvent>> sender;

    public EventPublisher(EventGridPublisherClient<EventGridEvent> client) {
        this(client::sendEvents);
    }

    public EventPublisher(Consumer<List<EventGridEvent>> sender) {
        this.sender = sender;
    }

    public void publish(List<CustomEvent> customEvents) {
        if (customEvents == null || customEvents.isEmpty()) {
            throw new IllegalArgumentException("At least one custom event is required");
        }
        sender.accept(customEvents.stream().map(EventPublisher::toEventGridEvent).toList());
    }

    static EventGridEvent toEventGridEvent(CustomEvent event) {
        return new EventGridEvent(
            event.subject(),
            event.type(),
            BinaryData.fromObject(event.data()),
            event.dataVersion());
    }
}
